🧭 Zadanie końcowe: przygotowanie danych do ML

1️⃣ Wczytanie i eksploracja danych
Wczytaj zbiór danych (pd.read_csv(...))

Sprawdź podstawowe informacje (head, info, describe)

Sprawdź typy danych i liczność klas (value_counts())

2️⃣ Czyszczenie danych
Zamień wartości ? na NaN

Sprawdź i obsłuż brakujące dane (isnull(), dropna(), fillna())

Usuń duplikaty (drop_duplicates())

3️⃣ Analiza kolumn
Podziel dane na kategoryczne i liczbowe

Sprawdź unikalne wartości, rozkład

Wstępna analiza statystyczna (opcjonalnie wizualizacje)

4️⃣ Optymalizacja typów danych
Zamień typy object na category, float64 na float32 itd.

Zmierz zużycie pamięci przed i po (memory_usage(deep=True))

5️⃣ Kodowanie zmiennych kategorycznych
Zastosuj get_dummies() lub LabelEncoder() w zależności od typu danych

6️⃣ Skalowanie danych numerycznych
Użyj StandardScaler, MinMaxScaler lub RobustScaler dla kolumn liczbowych

7️⃣ Podział na cechy i target
Oddziel dane wejściowe X od etykiety y

8️⃣ (Opcjonalnie) Balansowanie danych
Sprawdź rozkład klas w y

Jeśli potrzeba – zastosuj SMOTE lub inne metody balansowania

9️⃣ (Bonus) Stwórz prosty model
Podziel dane na train/test (train_test_split)

Wytrenuj klasyfikator (np. RandomForestClassifier)

Oceń wyniki (accuracy, classification_report, confusion_matrix)

In [24]:
import pandas as pd
import numpy as np

In [25]:
df = pd.read_csv('adult.csv')
df.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

In [26]:
df.describe(include='all')

age workclass        fnlwgt education  education.num  \
count   32561.000000     32561  3.256100e+04     32561   32561.000000   
unique           NaN         9           NaN        16            NaN   
top              NaN   Private           NaN   HS-grad            NaN   
freq             NaN     22696           NaN     10501            NaN   
mean       38.581647       NaN  1.897784e+05       NaN      10.080679   
std        13.640433       NaN  1.055500e+05       NaN       2.572720   
min        17.000000       NaN  1.228500e+04       NaN       1.000000   
25%        28.000000       NaN  1.178270e+05       NaN       9.000000   
50%        37.000000       NaN  1.783560e+05       NaN      10.000000   
75%        48.000000       NaN  2.370510e+05       NaN      12.000000   
max        90.000000       NaN  1.484705e+06       NaN      16.000000   

            marital.status      occupation relationship   race    sex  \
count                32561           32561        32561  32561  32561   
unique                   7              15            6      5      2   
top     Married-civ-spouse  Prof-specialty      Husband  White   Male   
freq                 14976            4140        13193  27816  21790   
mean                   NaN             NaN          NaN    NaN    NaN   
std                    NaN             NaN          NaN    NaN    NaN   
min                    NaN             NaN          NaN    NaN    NaN   
25%                    NaN             NaN          NaN    NaN    NaN   
50%                    NaN             NaN          NaN    NaN    NaN   
75%                    NaN             NaN          NaN    NaN    NaN   
max                    NaN             NaN          NaN    NaN    NaN   

        capital.gain  capital.loss  hours.per.week native.country income  
count   32561.000000  32561.000000    32561.000000          32561  32561  
unique           NaN           NaN             NaN             42      2  
top              NaN           NaN             NaN  United-States  <=50K  
freq             NaN           NaN             NaN          29170  24720  
mean     1077.648844     87.303830       40.437456            NaN    NaN  
std      7385.292085    402.960219       12.347429            NaN    NaN  
min         0.000000      0.000000        1.000000            NaN    NaN  
25%         0.000000      0.000000       40.000000            NaN    NaN  
50%         0.000000      0.000000       40.000000            NaN    NaN  
75%         0.000000      0.000000       45.000000            NaN    NaN  
max     99999.000000   4356.000000       99.000000            NaN    NaN

In [27]:
df.info(verbose=True, show_counts=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   age             int64 
 1   workclass       object
 2   fnlwgt          int64 
 3   education       object
 4   education.num   int64 
 5   marital.status  object
 6   occupation      object
 7   relationship    object
 8   race            object
 9   sex             object
 10  capital.gain    int64 
 11  capital.loss    int64 
 12  hours.per.week  int64 
 13  native.country  object
 14  income          object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


---

In [28]:
for col in df.select_dtypes(include=['object']).columns.tolist():
    print(f"{col}: {df[f'{col}'].unique()} | {len(df[f'{col}'].unique())}") # '?' is the issue

cols_del = df.columns[df.isin(values=['?', '', ' ']).any()] # Columns that has wrong values in there
print(f"Cols, where suspended values append: {cols_del.tolist()}")


workclass: ['?' 'Private' 'State-gov' 'Federal-gov' 'Self-emp-not-inc' 'Self-emp-inc'
 'Local-gov' 'Without-pay' 'Never-worked'] | 9
education: ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool'] | 16
marital.status: ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse'] | 7
occupation: ['?' 'Exec-managerial' 'Machine-op-inspct' 'Prof-specialty'
 'Other-service' 'Adm-clerical' 'Craft-repair' 'Transport-moving'
 'Handlers-cleaners' 'Sales' 'Farming-fishing' 'Tech-support'
 'Protective-serv' 'Armed-Forces' 'Priv-house-serv'] | 15
relationship: ['Not-in-family' 'Unmarried' 'Own-child' 'Other-relative' 'Husband' 'Wife'] | 6
race: ['White' 'Black' 'Asian-Pac-Islander' 'Other' 'Amer-Indian-Eskimo'] | 5
sex: ['Female' 'Male'] | 2
native.country: ['United-States' '?' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'Ind

In [29]:
print(f"indexes of duplicates{df[df.duplicated(keep=False)].index}. Length: {len(df[df.duplicated(keep=False)].index)}")

age_duplicates = df[df.duplicated(keep=False)].age.unique().tolist()
df[df.duplicated(keep=False)].loc[df['age'] == age_duplicates[1]] # Example of duplicated rows

indexes of duplicatesIndex([ 6227,  7615,  7978,  8356,  8453,  8500,  8645,  9067,  9269,  9305,
       10307, 10357, 11121, 11261, 11777, 12202, 13260, 14346, 14634, 15603,
       17344, 17456, 18417, 18906, 19014, 19067, 19622, 19832, 20388, 20507,
       22595, 22783, 22934, 23276, 23660, 23720, 23827, 26531, 26738, 27133,
       28796, 29051, 29334, 29604, 31060, 32065, 32419],
      dtype='int64'). Length: 47


age workclass  fnlwgt     education  education.num marital.status  \
7615    19   Private  251579  Some-college             10  Never-married   
10307   19   Private  138153  Some-college             10  Never-married   
17456   19   Private  146679  Some-college             10  Never-married   
19622   19   Private   97261       HS-grad              9  Never-married   
20507   19   Private   97261       HS-grad              9  Never-married   
22783   19   Private  138153  Some-college             10  Never-married   
22934   19   Private  146679  Some-college             10  Never-married   
32065   19   Private  251579  Some-college             10  Never-married   

            occupation   relationship   race     sex  capital.gain  \
7615     Other-service      Own-child  White    Male             0   
10307     Adm-clerical      Own-child  White  Female             0   
17456  Exec-managerial      Own-child  Black    Male             0   
19622  Farming-fishing  Not-in-family  White    Male             0   
20507  Farming-fishing  Not-in-family  White    Male             0   
22783     Adm-clerical      Own-child  White  Female             0   
22934  Exec-managerial      Own-child  Black    Male             0   
32065    Other-service      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country income  
7615              0              14  United-States  <=50K  
10307             0              10  United-States  <=50K  
17456             0              30  United-States  <=50K  
19622             0              40  United-States  <=50K  
20507             0              40  United-States  <=50K  
22783             0              10  United-States  <=50K  
22934             0              30  United-States  <=50K  
32065             0              14  United-States  <=50K

In [30]:
b_nan = df.isna().sum().sum()
df.replace(to_replace=['?'],
           value=np.nan,
           inplace=True)
a_nan = df.isna().sum().sum()

print(f"Before nans: {b_nan}\nAfter nan: {a_nan}")

Before nans: 0
After nan: 4262


In [31]:
df_size_before = len(df)

In [32]:
b_dup = df.duplicated().any()
df.drop_duplicates(inplace=True)
a_dup = df.duplicated().any()

print(f"Before duplicates: {b_dup}\nAfter duplicates: {a_dup}")

Before duplicates: True
After duplicates: False


In [33]:
df.dropna(inplace=True)
a2_nan = df.isna().sum().sum()
print(f"After deletion: {a2_nan}")

After deletion: 0


In [34]:
df_size_after = len(df)
print(f"Size before deletion: {df_size_before} -> Size after deletion: {df_size_after}. The percentage of deleted data: {((df_size_before - df_size_after)/df_size_before)*100:.2f} %")

Size before deletion: 32561 -> Size after deletion: 30139. The percentage of deleted data: 7.44 %


---

In [35]:
obj_cols = df.select_dtypes(include=['object']).columns
obj_cols

Index(['workclass', 'education', 'marital.status', 'occupation',
       'relationship', 'race', 'sex', 'native.country', 'income'],
      dtype='object')

In [36]:
country_counts = df['native.country'].value_counts() # liczba wystapien na narodowosc
country_counts

native.country
United-States                 27487
Mexico                          606
Philippines                     188
Germany                         128
Puerto-Rico                     109
Canada                          107
El-Salvador                     100
India                           100
Cuba                             92
England                          86
Jamaica                          80
South                            71
Italy                            68
China                            68
Dominican-Republic               67
Vietnam                          64
Guatemala                        61
Japan                            59
Poland                           56
Columbia                         56
Iran                             42
Taiwan                           42
Haiti                            42
Portugal                         34
Nicaragua                        33
Peru                             30
Greece                           29
Ecuador      

In [37]:
rare_countries = country_counts[country_counts < 100].index
print(len(rare_countries))

df['native.country'] = df['native.country'].replace(to_replace=rare_countries, value='Other')
df['native.country'].unique()

33


array(['United-States', 'Mexico', 'Other', 'India', 'Philippines',
       'Canada', 'Puerto-Rico', 'Germany', 'El-Salvador'], dtype=object)

In [38]:
df['income'] = df['income'].map({'<=50K': 0, '>50K': 1})
X = df.drop('income', axis=1)
y = df['income']

In [39]:
from sklearn.preprocessing import LabelEncoder

linear_model = False # Modele liniowe lub drzewiaste
swap_to_category = True # Zmiana typu danych na category
enc_cols = ['sex', 'workclass', 'education', 'marital.status', 'occupation', 'relationship', 'race', 'native.country']

if linear_model:
    X = pd.get_dummies(X, columns=enc_cols)
else:
    # X['sex'] = X['sex'].astype(dtype='category')
    # df['income'] = df['income'].astype(dtype='category')
    for col in enc_cols:
        X[col] = LabelEncoder().fit_transform(X[col])
        if swap_to_category:
            X[col] = X[col].astype(dtype='category')
    
X.head(5)

age workclass  fnlwgt education  education.num marital.status occupation  \
1   82         2  132870        11              9              6          3   
3   54         2  140359         5              4              0          6   
4   41         2  264663        15             10              5          9   
5   34         2  216864        11              9              0          7   
6   38         2  150601         0              6              5          0   

  relationship race sex  capital.gain  capital.loss  hours.per.week  \
1            1    4   0             0          4356              18   
3            4    4   0             0          3900              40   
4            3    4   0             0          3900              40   
5            4    4   0             0          3770              45   
6            4    4   1             0          3770              40   

  native.country  
1              8  
3              8  
4              8  
5              8  
6              8

In [40]:
int_cols = df.select_dtypes(include=['int64']).columns
for col in int_cols:
    df[col] = pd.to_numeric(df[col], downcast='integer')

df.dtypes

age                 int8
workclass         object
fnlwgt             int32
education         object
education.num       int8
marital.status    object
occupation        object
relationship      object
race              object
sex               object
capital.gain       int32
capital.loss       int16
hours.per.week      int8
native.country    object
income              int8
dtype: object

In [41]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

for int_col in int_cols:
    print(f"{int_col} | \n{df[int_col].agg(['min', 'max'])}\n")

int_col_scale = ['age', 'fnlwgt', 'capital.gain', 'capital.loss', 'education.num', 'hours.per.week']
X[int_col_scale] = scaler.fit_transform(X[int_col_scale])
X.head()

age | 
min    17
max    90
Name: age, dtype: int8

fnlwgt | 
min      13769
max    1484705
Name: fnlwgt, dtype: int64

education.num | 
min     1
max    16
Name: education.num, dtype: int8

capital.gain | 
min        0
max    99999
Name: capital.gain, dtype: int64

capital.loss | 
min       0
max    4356
Name: capital.loss, dtype: int16

hours.per.week | 
min     1
max    99
Name: hours.per.week, dtype: int8

income | 
min    0
max    1
Name: income, dtype: int8



age workclass    fnlwgt education  education.num marital.status  \
1  3.317157         2 -0.538773        11      -0.440434              6   
3  1.184832         2 -0.467892         5      -2.402221              0   
4  0.194824         2  0.708595        15      -0.048076              5   
5 -0.338257         2  0.256197        11      -0.440434              0   
6 -0.033639         2 -0.370956         0      -1.617506              5   

  occupation relationship race sex  capital.gain  capital.loss  \
1          3            1    4   0     -0.147502     10.551814   
3          6            4    4   0     -0.147502      9.424325   
4          9            3    4   0     -0.147502      9.424325   
5          7            4    4   0     -0.147502      9.102892   
6          0            4    4   1     -0.147502      9.102892   

   hours.per.week native.country  
1       -1.914647              8  
3       -0.078031              8  
4       -0.078031              8  
5        0.339381              8  
6       -0.078031              8

---

In [42]:
print(f"1: {len(y.loc[y == 1])} |  0: {len(y.loc[y == 0])}") # inbalance

1: 7506 |  0: 22633


In [43]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

X_resampled, y_resampled = smote.fit_resample(X, y)

print(f"1: {len(y_resampled.loc[y_resampled == 1])} |  0: {len(y_resampled.loc[y_resampled == 0])}") # Balanced

1: 22633 |  0: 22633


---

In [44]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, stratify=y)

In [45]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

if linear_model:
    model = LogisticRegression(max_iter= 1000, random_state=42)
else:
    model = DecisionTreeClassifier(random_state=42)

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [46]:
from sklearn.metrics import accuracy_score

acc = accuracy_score(y_true=y_test, y_pred=y_pred)
print(f"Accuracy: {acc:.2f}")

Accuracy: 0.81


Accuracy:

linear_model: 85%

tree_model: 81%

Imo okey result, in the future will be better